codalab/worker/docker_image_manager.py

```diff
@@ -218,12 +224,34 @@ def download():
                             'message'
                         ] = "Can't download image: {}".format(ex)
 
-                self._downloading.add_if_new(image_spec, threading.Thread(target=download, args=[]))
+                # Check docker image size before pulling from Docker Hub.
+                # Do not download images larger than self._max_image_size
+                image_size_bytes = docker_utils.get_image_size_without_pulling(image_spec)
+                if image_size_bytes and image_size_bytes < self._max_image_size:
+                    self._downloading.add_if_new(
+                        image_spec, threading.Thread(target=download, args=[])
+                    )
+                    return ImageAvailabilityState(
+                        digest=None,
+                        stage=DependencyStage.DOWNLOADING,
+                        message=self._downloading[image_spec]['status'],
+                    )
+                elif not image_size_bytes:
```

Comment:
Put this first (`image_size_bytes is None`), so that you don't need the `image_size_bytes` condition above.